nowcasting_dataset/data_sources/data_source.py

```diff
@@ -112,13 +121,90 @@ def open(self):
         """Open the data source, if necessary.
 
         Called from each worker process.  Useful for data sources where the
-        underlying data source cannot be forked (like Zarr on GCP!).
+        underlying data source cannot be forked (like Zarr).
+
+        Data sources which can be forked safely should call open() from __init__().
+        """
+        pass
+
+    def check_input_paths_exist(self) -> None:
+        """Check any input paths exist.  Raise FileNotFoundError if not.
 
-        Data sources which can be forked safely should call open()
-        from __init__().
+        Can be overridden by child classes.
         """
         pass
 
+    def create_batches(
+        self,
+        spatial_and_temporal_locations_of_each_example: pd.DataFrame,
+        idx_of_first_batch: int,
+        batch_size: int,
+        dst_path: Path,
+        local_temp_path: Path,
+        upload_every_n_batches: int,
+    ) -> None:
+        """Create multiple batches and save them to disk.
+
+        Safe to call from worker processes.
+
+        Args:
+          spatial_and_temporal_locations_of_each_example: A DataFrame where each row specifies
+            the spatial and temporal location of an example.  The number of rows must be
+            an exact multiple of `batch_size`.
+            Columns are: t0_datetime_UTC, x_center_OSGB, y_center_OSGB.
```

Comment:
nice column names, worth an assert at the start to check this is what comes in

Reply:
That's a great idea!  Done!